packages/driver/src/cypress/command_queue.ts

```diff
@@ -0,0 +1,400 @@
+
+interface Command {
+  get(key: string): any
+  get(): any
+  set(key: string, value: any): any
+  set(options: any): any
+  attributes: object
+  finishLogs(): void
+}
+
+const __stackReplacementMarker = (fn, ctx, args) => {
+  return fn.apply(ctx, args)
+}
+
+const commandRunningFailed = (Cypress, state, err) => {
+  // allow for our own custom onFail function
+  if (err.onFail) {
+    err.onFail(err)
+
+    // clean up this onFail callback after it's been called
+    delete err.onFail
+
+    return
+  }
+
+  const current = state('current')
+
+  return Cypress.log({
+    end: true,
+    snapshot: true,
+    error: err,
+    consoleProps () {
+      if (!current) return
+
+      const obj = {}
+      const prev = current.get('prev')
+
+      // if type isnt parent then we know its dual or child
+      // and we can add Applied To if there is a prev command
+      // and it is a parent
+      if (current.get('type') !== 'parent' && prev) {
+        const ret = $dom.isElement(prev.get('subject')) ?
+          $dom.getElements(prev.get('subject'))
+          :
+          prev.get('subject')
+
+        obj['Applied To'] = ret
+
+        return obj
+      }
+
+      return
+    },
+  })
+}
+
+export const create = (state, timeouts, stability, cleanup, fail, isCy) => {
+  const queue = createQueue()
+
+  const get = queue.get.bind(queue)
+  const slice = queue.slice.bind(queue)
+  const at = queue.at.bind(queue)
+  const reset = queue.reset.bind(queue)
+  const clear = queue.clear.bind(queue)
+  const stop = queue.stop.bind(queue)
+
+  const logs = (filter) => {
+    let logs = _.flatten(_.invokeMap(queue.get(), 'get', 'logs'))
+
+    if (filter) {
+      const matchesFilter = _.matches(filter)
+
+      logs = _.filter(logs, (log) => {
+        return matchesFilter(log.get())
+      })
+    }
+
+    return logs
+  }
+
+  const names = () => {
+    return _.invokeMap(queue.get(), 'get', 'name')
+  }
+
+  const add = (command) => {
+    queue.add(command)
+  }
+
+  const insert = (index: number, command: Command) => {
+    queue.insert(index, command)
+
+    const prev = at(index - 1) as Command
+    const next = at(index + 1) as Command
+
+    if (prev) {
+      prev.set('next', command)
+      command.set('prev', prev)
+    }
+
+    if (next) {
+      next.set('prev', command)
+      command.set('next', next)
+    }
+
+    return command
+  }
+
+  const find = (attrs) => {
+    const matchesAttrs = _.matches(attrs)
+
+    return _.find(queue.get(), (command: Command) => {
+      return matchesAttrs(command.attributes)
+    })
+  }
+
+  const runCommand = (command: Command) => {
+    // bail here prior to creating a new promise
+    // because we could have stopped / canceled
+    // prior to ever making it through our first
+    // command
+    if (queue.stopped) {
+      return
+    }
+
+    state('current', command)
+    state('chainerId', command.get('chainerId'))
+
+    return stability.whenStable(() => {
```

Comment:
should we start using async/await for stuff like this? all the nested .then callbacks drive me crazy